### PR TITLE
expression: Add test cases to cover the cases when invalid int value is casted as TIME (#18653)

### LINKED ISSUE #18653: CAST function cannot handle invalid value to TIME
### 1. Minimal reproduce step (Required)

```` MySQL
CAST( -17783 AS TIME );
````

### 2. What did you expect to see? (Required)

```` MySQL
mysql> select  CAST( -17783 AS TIME ) ;
+------------------------+
| CAST( -17783 AS TIME ) |
+------------------------+
| NULL                   |
+------------------------+
1 row in set, 1 warning (0.00 sec)
````

### 3. What did you see instead (Required)

```` MySQL
mysql> select CAST( -17783 AS TIME );
ERROR 1292 (22007): Incorrect time value: '17783'
````

### 4. Affected version (Required)

````
commit 9e5e2a933e8f311899d8594f3310c59770092fa8 (HEAD -> master, origin/master, origin/HEAD)
Author: Christina Fritz <67818510+tina77fritz@users.noreply.github.com>
Date:   Wed Jul 15 17:46:11 2020 +080
````

### PR DESCRIPTION
**Pull request by xuhui-lu**
### What problem does this PR solve?

Issue Number: close #18653 

Problem Summary:
Current TIDB version didn't handle the error when invalid error is casted as TIME. This is fixed in another PR, thus this PR will just add more test cases to cover that.

### What is changed and how it works?

What's Changed:

More Test cases are added.

How it Works:

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- Make Cast Function returns Null when invalid int value is casted as TIME

### CODE REVIEW

File: expression/errors.go [hunk 1]
```diff
@@ -67,6 +67,20 @@ func handleInvalidTimeError(ctx sessionctx.Context, err error) error {
...
 
+// handleInvalidDurationError reports error or warning depend on the context.
+func handleInvalidDurationError(ctx sessionctx.Context, err error) error {
+	if err == nil || !(types.ErrWrongValue.Equal(err) || types.ErrWrongValueForType.Equal(err) ||
+		types.ErrTruncatedWrongVal.Equal(err)) {
+		return err
+	}
+	sc := ctx.GetSessionVars().StmtCtx
+	err = sc.HandleTruncate(err)
+	if ctx.GetSessionVars().StrictSQLMode && (sc.InInsertStmt || sc.InUpdateStmt || sc.InDeleteStmt) {
```
Comment (wjhuang2016) [inline expression/errors.go:78]: I'm not sure if this line is necessary. The error should be handled by HandleTruncate().

**Review by wjhuang2016 — COMMENTED** [review]
Please add some unit tests.

File: expression/builtin_cast.go [hunk 2]
```diff
@@ -604,7 +604,7 @@ func (b *builtinCastIntAsDurationSig) evalDuration(row chunk.Row) (res types.Dur
 		if types.ErrOverflow.Equal(err) {
 			err = b.ctx.GetSessionVars().StmtCtx.HandleOverflow(err, err)
 		}
-		return res, true, err
+		return res, true, handleInvalidDurationError(b.ctx, err)
```
Comment (SunRunAway) [inline expression/builtin_cast.go:607]: `builtinCastIntAsDurationSig.vecEvalDuration` should also be fixed.

**Review by wjhuang2016 — COMMENTED** [review]
```
        dur, err := types.NumberToDuration(val, int8(b.tp.Decimal))
        if err != nil {
-               if types.ErrOverflow.Equal(err) {
-                       err = b.ctx.GetSessionVars().StmtCtx.HandleOverflow(err, err)
-               }
+               err = handleInvalidTimeError(b.ctx, err)
                return res, true, err
        }
        return dur, false, err
```
@xuhui-lu  I think this is enough. The hard part of this issue is that, you need to modify TiKV part as well.

**Review by wjhuang2016 — COMMENTED** [review]
LGTM

**Review by wjhuang2016 — COMMENTED** [review]
Please modify the PR's title and description.

**Comment by b41sh** [comment]
The modified code seems lost. @xuhui-lu

**Comment by xuhui-lu** [comment]
> The modified code seems lost. @xuhui-lu

The error handling is added by another PR. Thus I deleted the error handling part and keep the test cases.

**Comment by b41sh** [comment]
> > The modified code seems lost. @xuhui-lu
> 
> The error handling is added by another PR. Thus I deleted the error handling part and keep the test cases.

ok. I see

**Comment by wjhuang2016** [comment]
@b41sh PTAL

**Comment by zz-jason** [comment]
I'm going to close this PR since the original issue has been fixed.

**Pull request closed without merging**